Lib/asyncio/tasks.py

```diff
@@ -75,6 +76,31 @@ def _set_task_name(task, name):
             set_name(name)
 
 
+@contextlib.contextmanager
+def shield_scope():
+    shield = object()  # just for making sure shields closed in order
+    # if __enter__ and __exit__ called directly
+    task = current_task()
+    task.shields.append(shield)
+    try:
+        yield
+    finally:
+        if task.shields.pop() is not shield:
+            raise RuntimeError(
+                f"Unbalanced enter/exit of shield scopes for task {task!r}"
+            )
+
+
+def shielded(func):
+    """Decorator to shield the call of a async function or method,
+    wrapping it in a shield_scope()."""
+    @functools.wraps(func)
+    async def _shielded(*args, **kwargs):
+        with shield_scope():
+            return await func(*args, **kwargs)
+    return _shielded
```

Comment:
Do we absolutely, truly need two ways to do it (a decorator and a context manager)? I'd start with just the context manager.